pandas/core/internals/blocks.py

```diff
@@ -1967,6 +1968,57 @@ def shift(self, periods, axis=0):
                                            placement=self.mgr_locs,
                                            ndim=self.ndim)]
 
+    def where(self, other, cond, align=True, errors='raise',
+              try_cast=False, axis=0, transpose=False):
+        if isinstance(other, (ABCIndexClass, ABCSeries)):
+            other = other.array
+
+        if isinstance(cond, ABCDataFrame):
+            assert cond.shape[1] == 1
+            cond = cond.iloc[:, 0].array
+
+        if isinstance(other, ABCDataFrame):
+            assert other.shape[1] == 1
+            other = other.iloc[:, 0].array
+
+        if isinstance(cond, (ABCIndexClass, ABCSeries)):
+            cond = cond.array
+
+        if lib.is_scalar(other) and isna(other):
+            # The default `other` for Series / Frame is np.nan
+            # we want to replace that with the correct NA value
+            # for the type
+            other = self.dtype.na_value
+
+        if is_sparse(self.values):
```

Comment:
Without this, we fail in the

```python
            result = self._holder._from_sequence(
                 np.where(cond, self.values, other),
                 dtype=dtype,
```

since the `where` may change the dtype, if NaN is introduced.


Implementing `SparseArray.__setitem__` would allow us to remove this block.

Reply:
this should be an overriding method in Sparse then, not here

Reply:
We don't have a SparseBlock anymore. I can add one back if you want, but I figured it'd be easier not to since implementing `SparseArray.__setitem__` will remove the need for this, and we'd just have to remove SparseBlock again.

Reply:
this is pretty hacky. This was why we had originally a ``.get_values()`` methon on Sparse to do things like this. We need something to give back the underlying type of the object, which is useful for Categorical as well. Would rather create a generalized soln than hack it like this.

Reply:
Actually, we don't need this. I think we can just re-infer the dtype from the output of `np.where`.

Reply:
so is this changing?

Reply:
Changing from master? Yes, in the sense that it'll return a SparseArray. But it still densifies when `np.where` is called.

If you mean "is this changing in the future", yes it'll be removed when `SparseArray.__setitem__` is implemented.

Reply:
oh ok, can you add a TODO comment